stolon/src/lib.rs

```diff
@@ -0,0 +1,89 @@
+use std::{net::SocketAddr, sync::Arc};
+
+use tokio::net::TcpStream;
+use tokio_socks::tcp::Socks5Stream;
+use tower::{Service, ServiceExt};
+
+pub use zebra_chain::{
+    parameters::Network, serialization::ZcashDeserialize, transaction::Transaction,
+};
+
+pub type BoxError = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+// Tor's IsolateSOCKSAuth option (on by default) prevents sharing circuits
+// for streams for which different SOCKS authentication was provided.
+
+const SOCKS_USERNAME: &str = "stolon";
+fn socks_password() -> String {
+    use rand::distributions::{Alphanumeric, Distribution};
+    Alphanumeric
+        .sample_iter(rand::thread_rng())
+        .take(16)
+        .collect()
+}
+
+/// Send a transaction over Tor to the Zcash network, using DNS seeders to find peers.
+#[tracing::instrument(skip(transaction))]
+pub async fn send_transaction(
+    tor_addr: SocketAddr,
+    network: Network,
+    transaction: Arc<Transaction>,
+) -> Result<(), BoxError> {
+    // We want to ensure that we get a different peer from the DNS seeder for
+    // each call, so that fanout will send multiple times to different peers
+    // rather than multiple times to the same peer. However, Tor only sends a
+    // single IP address per RESOLVED cell, and caches DNS lookups. As a
+    // workaround, instead of looking up an IP and then connecting, try
+    // connecting directly to the DNS seeder, so that DNS resolution is
+    // performed by the exit. Since we use a new circuit for each call, we hit a
+    // different exit each time.
+    let seed_domain = match network {
+        Network::Mainnet => "mainnet.seeder.zfnd.org:8233",
+        Network::Testnet => "testnet.seeder.zfnd.org:18233",
```

Comment:
Should we choose between multiple DNS seeders at random?
Should we fallback to another seeder if the one we choose is down?

Reply:
If we could resolve DNS addresses to multiple IPs and collect the resulting sets (like we can do with https://doc.rust-lang.org/nightly/std/net/trait.ToSocketAddrs.html), I think this would be easier (we could assemble a list of candidates, and sample within that list). Unfortunately, because we do DNS resolution on a per-connection basis, we can't look up multiple seeders and combine results, so we'd have to pick one per connection attempt.

I'm not sure that we would be able to tell that an error occurred because the seeder is down, rather than for some other reason (e.g., transit through a hostile network that can and does modify data).

We could try to pick one domain per `send_transaction` call, e.g., selecting between the ZF seeder and the ECC seeder with 50%/50% probability.  I don't want to have a large list of DNS seeders, because I think it would be preferable to have a small number of well-maintained infrastructure components rather than a larger number of less well-maintained ones.

Reply:
Yeah, Tor doesn't let you choose between multiple DNS addresses for the same connection.

If the goal is to reduce the probability of failure, then we should try to connect to each seeder once for every transaction. Then, if one seeder is down, the other seeder should handle the request.

For example, here is one possible fanout design:
1. Connect to the ZFND seeder
2. Connect to the ECC seeder
3. For the remaining fanouts, connect uniformly at random to either seeder

Reply:
From an implementation point of view I think it would be preferable if all of the `send_transaction` tasks could be fully independent, but if we force choosing the ZF / ECC seeders at least once there would have to be some communication.  WDYT of the idea of connecting uniformly at random to either the ECC or ZF seeder in each `send_transaction` call?